src/java/org/apache/cassandra/db/memtable/Memtable.java

```diff
@@ -0,0 +1,389 @@
+    long getOperations();
+
+    /** Minimum timestamp of all stored data */
+    long getMinTimestamp();
+
+    /**
+     * The table's definition metadata.
+     *
+     * Note that this tracks the current state of the table and is not necessarily the same as what was used to create
+     * the memtable.
+     */
+    TableMetadata metadata();
+
+
+    // Memory usage tracking
+
+    /**
+     * This memtable's used memory.
+     *
+     * This tracks on- and off-heap memory, as well as the ratio to the total permitted memtable memory.
+     */
+    MemoryUsage getMemoryUsage();
+
+    /**
+     * Add this memtable's used memory to the given usage object. Used mainly to aggregate the usage data between a
+     * table and its table-based indexes.
+     */
+    void addMemoryUsageTo(MemoryUsage usage);
+
+    class MemoryUsage
+    {
+        /** On-heap memory used in bytes */
+        public long ownsOnHeap = 0;
+        /** Off-heap memory used in bytes */
+        public long ownsOffHeap = 0;
+        /** On-heap memory as ratio to permitted memtable space */
+        public float ownershipRatioOnHeap = 0.0f;
+        /** Off-heap memory as ratio to permitted memtable space */
+        public float ownershipRatioOffHeap = 0.0f;
+        /** The memtable memory pool that this memtable is using */
+        public final MemtablePool memoryPool;
+
+        public MemoryUsage(MemtablePool memoryPool)
+        {
+            this.memoryPool = memoryPool;
+        }
+
+        public String toString()
+        {
+            return String.format("%s (%.0f%%) on-heap, %s (%.0f%%) off-heap",
+                                 FBUtilities.prettyPrintMemory(ownsOnHeap),
+                                 ownershipRatioOnHeap * 100,
+                                 FBUtilities.prettyPrintMemory(ownsOffHeap),
+                                 ownershipRatioOffHeap * 100);
+        }
+    }
+
+    /**
+     * Adjust the used on-heap space by the given size (e.g. to reflect memory used by a non-table-based index).
+     * This operation may block until enough memory is available in the memory pool.
+     *
+     * @param additionalSpace the number of allocated bytes
+     * @param opGroup write operation group, used to permit the operation to complete if it is needed to complete a
+     *                flush to free space.
+     */
+    void allocateExtraOnHeap(long additionalSpace, OpOrder.Group opGroup);
+
+    /**
+     * Adjust the used off-heap space by the given size (e.g. to reflect memory used by a non-table-based index).
+     * This operation may block until enough memory is available in the memory pool.
+     *
+     * @param additionalSpace the number of allocated bytes
+     * @param opGroup write operation group, used to permit the operation to complete if it is needed to complete a
+     *                flush to free space.
+     */
+    void allocateExtraOffHeap(long additionalSpace, OpOrder.Group opGroup);
```

Comment:
IIUC `allocate` here does not mean any allocation, but is a memory backpressure mechanism?
e.g. - it is called post factum, after actual memory has been used by some external compoennt (e.g. index), and there is no allocation happening, no memory is actually allocated and returned to be used by anyone?

Is that a C* convention to use the `allocate` term for such purpose? It got me baffled until I checked the implementation.
Does it make sense to rename the API to sth along: `markExtraOnHeapUsed` ?

Reply:
Renamed.